File: internal/domain/transaction.go
package domain

import (
	"time"

	"gorm.io/gorm"
)

const (
	TransactionTypeIncome   = "income"
	TransactionTypeExpense  = "expense"
	TransactionTypeTransfer = "transfer"
)

type Transaction struct {
	ID                  string         `gorm:"type:uuid;default:gen_random_uuid();primaryKey" json:"id"`
	UserID              string         `gorm:"type:uuid;not null;index" json:"user_id"`
	Type                string         `gorm:"type:varchar(50);not null" json:"type"`
	Amount              float64        `gorm:"type:numeric;not null" json:"amount"`
	CategoryID          *string        `gorm:"type:uuid;index" json:"category_id"`
	SourceWalletID      *string        `gorm:"type:uuid;index" json:"source_wallet_id"`
	DestinationWalletID *string        `gorm:"type:uuid;index" json:"destination_wallet_id"`
	ReferenceNo         *string        `gorm:"type:varchar(255)" json:"reference_no"`
	Note                *string        `gorm:"type:text" json:"note"`
	AttachmentURL       *string        `gorm:"type:text" json:"attachment_url"`
	TransactionDate     time.Time      `gorm:"not null" json:"transaction_date"`
	CreatedAt           time.Time      `gorm:"not null" json:"created_at"`
	UpdatedAt           *time.Time     `json:"updated_at"`
	DeletedAt           gorm.DeletedAt `gorm:"index" json:"-"`
}

File: money-manager-frontend/src/types/transaction.ts
import { z } from "zod";
import { TransactionType } from "./index";

export interface TransactionFilterParams {
  page?: number;
  limit?: number;
  search?: string;
  type?: TransactionType;
  category_id?: string;
  wallet_id?: string;
  min_amount?: number;
  max_amount?: number;
  month?: number;
  year?: number;
  start_date?: string;
  end_date?: string;
  sort?: "newest" | "oldest" | "highest" | "lowest";
}

export const transactionSchema = z.object({
  type: z.enum(["income", "expense", "transfer"]),
  amount: z.coerce.number().gt(0, "Amount must be greater than 0"),
  category_id: z.string().uuid("Please select a category").optional().nullable(),
  source_wallet_id: z.string().uuid("Please select a source wallet").optional().nullable(),
  destination_wallet_id: z.string().uuid("Please select a destination wallet").optional().nullable(),
  reference_no: z.string().optional().nullable(),
  note: z.string().optional().nullable(),
  attachment_url: z.string().url("Please enter a valid URL").optional().nullable().or(z.literal("")),
  transaction_date: z.string().min(1, "Date is required"),
}).refine((data) => {
  if (data.type === "income") {
    return !!data.destination_wallet_id;
  }
  if (data.type === "expense") {
    return !!data.source_wallet_id;
  }
  if (data.type === "transfer") {
    return !!data.source_wallet_id && !!data.destination_wallet_id && data.source_wallet_id !== data.destination_wallet_id;
  }
  return true;
}, {
  message: "Wallet selection is invalid for the transaction type",
  path: ["source_wallet_id"], // Default path
});

export type TransactionFormData = z.infer<typeof transactionSchema>;
